stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -639,20 +642,33 @@ public void onClose(Status status, Metadata trailers) {
      * Waits until there is a Runnable, then executes it and all queued Runnables after it.
      */
     public void waitAndDrain() throws InterruptedException {
-      Runnable runnable = queue.take();
-      while (runnable != null) {
+      Runnable runnable = poll();
+      if (runnable == null) {
+        waiter = Thread.currentThread();
+        try {
+          while ((runnable = poll()) == null) {
+            LockSupport.park(this);
+          }
+        } finally {
+          waiter = null;
+        }
+      }
+      do {
         try {
           runnable.run();
         } catch (Throwable t) {
           log.log(Level.WARNING, "Runnable threw exception", t);
         }
-        runnable = queue.poll();
-      }
+      } while ((runnable = poll()) != null);
     }
 
     @Override
     public void execute(Runnable runnable) {
-      queue.add(runnable);
+      add(runnable);
+      final Thread waitingThread = waiter;
+      if (waitingThread != null) {
+        LockSupport.unpark(waitingThread);
```

Comment:
Why not just `LockSupport.unpark(waiter)`? If waiter is null, there is no effect.